internal/api/apiservice/rebuild.go

```diff
@@ -199,38 +193,105 @@ func RebuildPackage(ctx context.Context, req schema.RebuildPackageRequest, deps
+		return api.AsStatus(codes.InvalidArgument, errors.New("unsupported ecosystem"))
 	}
 	if err != nil {
-		return nil, api.AsStatus(codes.Internal, errors.Wrap(err, "rebuilding"))
+		return api.AsStatus(codes.Internal, errors.Wrap(err, "rebuilding"))
 	}
 	rb, up, err := verifier.SummarizeArtifacts(ctx, metadata, t, upstreamURI, hashes)
 	if err != nil {
-		return nil, api.AsStatus(codes.Internal, errors.Wrap(err, "comparing artifacts"))
+		return api.AsStatus(codes.Internal, errors.Wrap(err, "comparing artifacts"))
 	}
 	exactMatch := bytes.Equal(rb.Hash.Sum(nil), up.Hash.Sum(nil))
 	canonicalizedMatch := bytes.Equal(rb.CanonicalHash.Sum(nil), up.CanonicalHash.Sum(nil))
 	if !exactMatch && !canonicalizedMatch {
-		return nil, api.AsStatus(codes.FailedPrecondition, errors.Wrap(err, "rebuild content mismatch"))
+		return api.AsStatus(codes.FailedPrecondition, errors.Wrap(err, "rebuild content mismatch"))
 	}
-	input := rebuild.Input{Target: t, Strategy: manualStrategy}
-	eqStmt, buildStmt, err := verifier.CreateAttestations(ctx, input, strategy, id, rb, up, metadata, buildDefLoc)
+	input := rebuild.Input{Target: t, Strategy: strat.ManualStrategy}
+	eqStmt, buildStmt, err := verifier.CreateAttestations(ctx, input, strat.Strategy, id, rb, up, metadata, strat.buildDefLoc)
 	if err != nil {
-		return nil, api.AsStatus(codes.Internal, errors.Wrap(err, "creating attestations"))
+		return api.AsStatus(codes.Internal, errors.Wrap(err, "creating attestations"))
 	}
 	if err := a.PublishBundle(ctx, t, eqStmt, buildStmt); err != nil {
-		return nil, api.AsStatus(codes.Internal, errors.Wrap(err, "publishing bundle"))
+		return api.AsStatus(codes.Internal, errors.Wrap(err, "publishing bundle"))
+	}
+	return nil
+}
+
+func doRebuildPackage(ctx context.Context, req schema.RebuildPackageRequest, deps *RebuildPackageDeps, mux rebuild.RegistryMux, t rebuild.Target) (*StrategyAndOrigin, error) {
+	signer := verifier.InTotoEnvelopeSigner{EnvelopeSigner: deps.Signer}
+	a := verifier.Attestor{Store: deps.AttestationStore, Signer: signer, AllowOverwrite: deps.OverwriteAttestations}
+	if !deps.OverwriteAttestations {
+		if exists, err := a.BundleExists(ctx, t); err != nil {
+			return nil, api.AsStatus(codes.Internal, errors.Wrap(err, "checking existing bundle"))
+		} else if exists {
+			return nil, api.AsStatus(codes.AlreadyExists, errors.New("conflict with existing attestation bundle"))
+		}
+	}
+	strat, err := getStrategy(ctx, deps, t, req.StrategyFromRepo)
+	if err != nil {
+		return nil, api.AsStatus(codes.Internal, errors.Wrap(err, "getting strategy"))
+	}
+	err = buildAndAttest(ctx, deps, mux, a, t, *strat)
+	if err != nil {
+		return strat, api.AsStatus(codes.Internal, errors.Wrap(err, "executing rebuild"))
+	}
+	return strat, nil
+}
+
+func RebuildPackage(ctx context.Context, req schema.RebuildPackageRequest, deps *RebuildPackageDeps) (*schema.Verdict, error) {
+	t := rebuild.Target{Ecosystem: req.Ecosystem, Package: req.Package, Version: req.Version}
```

Comment:
Originally I did this logic at the top level because I wanted the error to be handled differently (an error populating the artifact should prevent firestore logging because theoretically firestore should use the artifact name when creating the record).

After pushing this logic inside doRebuildPackage, I made that specific error return an error from doRebuildPackage. Otherwise all other errors get pushed into the verdict being returned.